manager/manager.go

```diff
@@ -0,0 +1,359 @@
+package manager
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/docker/infrakit/discovery"
+	"github.com/docker/infrakit/leader"
+	rpc "github.com/docker/infrakit/rpc/group"
+	"github.com/docker/infrakit/spi/group"
+	"github.com/docker/infrakit/store"
+)
+
+// Backend is the admin / server interface
+type Backend interface {
+	group.Plugin
+
+	Start() (<-chan struct{}, error)
+	Stop()
+}
+
+// manager is the controller of all the plugins.  It is able to process multiple inputs
+// such as leadership changes and configuration changes and perform the necessary actions
+// to activate / deactivate plugins
+type manager struct {
+	group.Plugin
+
+	plugins  discovery.Plugins
+	leader   leader.Detector
+	snapshot store.Snapshot
+	isLeader bool
+	lock     sync.Mutex
+	stop     chan struct{}
+	running  chan struct{}
+
+	currentConfig GlobalSpec
+
+	backendName string
+	backendOps  chan<- backendOp
+}
+
+type backendOp struct {
+	n   string
+	f   func() error
+	err chan<- error
+}
+
+// NewManager returns the manager which depends on other services to coordinate and manage
+// the plugins in order to ensure the infrastructure state matches the user's spec.
+func NewManager(plugins discovery.Plugins, leader leader.Detector, snapshot store.Snapshot,
+	backendName string) (Backend, error) {
+
+	m := &manager{
+		plugins:  plugins,
+		leader:   leader,
+		snapshot: snapshot,
+	}
+
+	gp, err := m.proxyForGroupPlugin(backendName)
+	if err != nil {
+		return nil, err
+	}
+
+	m.Plugin = gp
+	return m, nil
+}
+
+// Start starts the manager.  It does not block. Instead read from the returned channel to block.
+func (m *manager) Start() (<-chan struct{}, error) {
+	m.lock.Lock()
+	if m.running != nil {
+		m.lock.Unlock()
```

Comment:
I'm a bit uneasy with the ad hoc locking starting to show up.

Possibly only worth dealing with later, but it would be nice to establish a straightforward convention for lock management.  It would be nice if we could strongly favor associating lock acquisition with function calls, like the pattern used elsewhere in this file:

```go
func ... {
	m.lock.Lock()
	defer m.lock.Unlock()
}
```

What do you think?